dot/state/epoch.go

```diff
@@ -365,3 +506,125 @@ func (s *EpochState) SkipVerify(header *types.Header) (bool, error) {
 
 	return false, nil
 }
+
+// StoreBABENextEpochData stores the types.NextEpochData under epoch and hash keys
+func (s *EpochState) StoreBABENextEpochData(epoch uint64, hash common.Hash, nextEpochData types.NextEpochData) {
+	s.nextEpochDataLock.Lock()
+	defer s.nextEpochDataLock.Unlock()
+
+	_, has := s.nextEpochData[epoch]
+	if !has {
+		s.nextEpochData[epoch] = make(map[common.Hash]types.NextEpochData)
+	}
+	s.nextEpochData[epoch][hash] = nextEpochData
+}
+
+// StoreBABENextConfigData stores the types.NextConfigData under epoch and hash keys
+func (s *EpochState) StoreBABENextConfigData(epoch uint64, hash common.Hash, nextConfigData types.NextConfigData) {
+	s.nextConfigDataLock.Lock()
+	defer s.nextConfigDataLock.Unlock()
+
+	_, has := s.nextConfigData[epoch]
+	if !has {
+		s.nextConfigData[epoch] = make(map[common.Hash]types.NextConfigData)
+	}
+	s.nextConfigData[epoch][hash] = nextConfigData
+}
+
+// FinalizeBABENextEpochData stores the right types.NextEpochData by
+// getting the set of hashes from the received epoch and for each hash
+// check if the header is in the database then it's been finalized and
+// thus we can also set the corresponding EpochData in the database
+func (s *EpochState) FinalizeBABENextEpochData(epoch uint64) error {
+	s.nextEpochDataLock.Lock()
+	defer s.nextEpochDataLock.Unlock()
+
+	finalizedNextEpochData, err := lookupForNextEpochPersistedHash(s.nextEpochData, s, epoch)
+	if err != nil {
+		return fmt.Errorf("cannot find next epoch data: %w", err)
+	}
+
+	ed, err := finalizedNextEpochData.ToEpochData()
+	if err != nil {
+		return fmt.Errorf("cannot transform epoch data: %w", err)
+	}
+
+	err = s.SetEpochData(epoch, ed)
+	if err != nil {
+		return fmt.Errorf("cannot set epoch data: %w", err)
+	}
+
+	// remove previous epochs from the memory
+	epochsToDelete := make([]uint64, 0)
```

Comment:
nit what's the point to assign it to a length of `0`?

Just do this instead?

```suggestion
	var epochsToDelete []uint64
```